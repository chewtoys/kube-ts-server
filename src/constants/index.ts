// HTTP
export const LIVENESS_CHECK_URL = '/health/liveness';
export const READINESS_CHECK_URL = '/health/readiness';
export const VERSION_CHECK_URL = '/version';

export const TRUST_PROXY = true;
export const EXPRESS_PORT = 3000;
export const EXPRESS_HOST = 'http://localhost';
export const CLIENT_URL = 'http://localhost:9000';
export const CLIENT_VERIFY_EMAIL_URL = 'http://localhost:9000/verify'; // @note: this is client url
export const CLIENT_VERIFY_TOKEN_QUERY_PARAM_NAME = 'token';
export const CLIENT_RESET_PASSWORD_URL = 'http://localhost:9000/reset-password'; // @note: this is client url
export const CLIENT_RESET_PASSWORD_TOKEN_QUERY_PARAM_NAME = 'token';

export const TOO_MANY_REQUEST_MESSAGE =
  'Too many accounts created from this IP, please try again after an 10 minutes';

// AUTH
export const JWT_EXPIRES_IN_MS = 86400000; // 24hrs
export const JWT_ALGORITM = 'HS256';
export const AUTH_SCHEME_NAME = 'Bearer';
export const AUTH_HEADER_NAME = 'authorization';
export const AUTH_QUERY_PARAM_NAME = 'auth_token';
export const AUTH_BODY_FIELD_NAME = 'auth_token';
export const ACCOUNT_LOCKOUT_TIME_IN_MINUTES = 10;
export const MAX_NUMBER_OF_FAILED_LOGIN_ATTEMPTS = 5;
export const DEFAULT_RESET_PASSWORD_TIME_IN_MINUTES = 60;

// MODEL
export const UUID_LENGTH = 36;
export const VARCHAR_LENGTH = 255;
export const TEXT_LENGTH = 65535;

// MAIL
export const SMTP_API_KEY = 'xxxxx';
export const SMTP_DOMAIN = 'example.com';
export const SMTP_FROM = 'noreply@example.com';
export const SMTP_HOST = 'smtp.example.com';
export const SMTP_IGNORE_TLS = false;
export const SMTP_REQUIRE_TLS = false;
export const SMTP_TEST_HOST = 'localhost';
export const SMTP_SECURE = false;
export const SMTP_PASS = 'password';
export const MAIL_PORT = 587;
export const SMTP_SERVICE = 'mailgun';
export const SMTP_TO = 'support@example.com';
export const SMTP_USER = 'user';
export const MAIL_REPO_TYPE = 'nodemailer';

// SUPPORTED LOCALES
export const DEFAULT_LOCALE = 'en';
export const SUPPORTED_LOCALES = ['en', 'pl'];
export const LOCALE_HEADER = 'accept-language';
export const LOCALE_QUERY_PARAM = 'lang';

// REGEXP
export const REGEXP_EMAIL = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

// minimum eight characters, at least one upper case letter, at least one lower case letter, one number and one special character:
// @credits: https://stackoverflow.com/questions/19605150/regex-for-password-must-contain-at-least-eight-characters-at-least-one-number-a
export const REGEXP_PASSWORD = /^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$/;

// DATETIME
export const DEFAULT_DATE_FORMATS = ['DD-MM-YYYY', 'YYYY-MM-DD'];

// note double backslashes to escape the chars used for regexp queries in mysql
export const QUERY_REGEX = `(\\?([a-z0-9$_\\.\\+!\\*\\'\\(\\),;:@&=-]|%[0-9a-f]{2})*)?`;

// @credits: https://stackoverflow.com/questions/136505/searching-for-uuids-in-text-with-regex
export const UUID_REGEX =
  '([0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}){1}';
