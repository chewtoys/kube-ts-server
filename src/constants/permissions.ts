export const USERS_GET_ITEM = 'users.getItem';
export const USERS_GET_ITEMS = 'users.getItems';
export const USERS_DELETE_ITEM = 'users.deleteItem';
export const USERS_DELETE_ITEMS = 'users.deleteItems';
export const USERS_UPDATE_ITEM = 'users.updateItem';
export const USERS_REPLACE_ITEM = 'users.replaceItem';
export const USERS_CREATE_ITEM = 'users.createItem';
export const USERS_ASSIGN_ROLE = 'users.assignUserRole';
export const USERS_REVOKE_ROLE = 'users.revokeUserRole';

export const usersPermissions = [
  USERS_GET_ITEM,
  USERS_GET_ITEMS,
  USERS_DELETE_ITEM,
  USERS_DELETE_ITEMS,
  USERS_UPDATE_ITEM,
  USERS_REPLACE_ITEM,
  USERS_CREATE_ITEM,
  USERS_ASSIGN_ROLE,
  USERS_REVOKE_ROLE,
];

export const ROLES_GET_ITEM = 'roles.getItem';
export const ROLES_GET_ITEMS = 'roles.getItems';
export const ROLES_DELETE_ITEM = 'roles.deleteItem';
export const ROLES_DELETE_ITEMS = 'roles.deleteItems';
export const ROLES_UPDATE_ITEM = 'roles.updateItem';
export const ROLES_REPLACE_ITEM = 'roles.replaceItem';
export const ROLES_CREATE_ITEM = 'roles.createItem';
export const ROLES_ASSIGN_PERMISSION = 'roles.assignRolePermission';
export const ROLES_REVOKE_PERMISSION = 'roles.revokeRolePermission';

export const rolesPermissions = [
  ROLES_GET_ITEM,
  ROLES_GET_ITEMS,
  ROLES_DELETE_ITEM,
  ROLES_DELETE_ITEMS,
  ROLES_UPDATE_ITEM,
  ROLES_REPLACE_ITEM,
  ROLES_CREATE_ITEM,
  ROLES_ASSIGN_PERMISSION,
  ROLES_REVOKE_PERMISSION,
];

export const PERMISSIONS_GET_ITEM = 'permissions.getItem';
export const PERMISSIONS_GET_ITEMS = 'permissions.getItems';
export const PERMISSIONS_DELETE_ITEM = 'permissions.deleteItem';
export const PERMISSIONS_DELETE_ITEMS = 'permissions.deleteItems';
export const PERMISSIONS_UPDATE_ITEM = 'permissions.updateItem';
export const PERMISSIONS_REPLACE_ITEM = 'permissions.replaceItem';
export const PERMISSIONS_CREATE_ITEM = 'permissions.createItem';

export const permissionsPermissions = [
  PERMISSIONS_GET_ITEM,
  PERMISSIONS_GET_ITEMS,
  PERMISSIONS_DELETE_ITEM,
  PERMISSIONS_DELETE_ITEMS,
  PERMISSIONS_UPDATE_ITEM,
  PERMISSIONS_REPLACE_ITEM,
  PERMISSIONS_CREATE_ITEM,
];

export const COURSES_GET_ITEM = 'courses.getItem';
export const COURSES_GET_ITEMS = 'courses.getItems';
export const COURSES_DELETE_ITEM = 'courses.deleteItem';
export const COURSES_DELETE_ITEMS = 'courses.deleteItems';
export const COURSES_UPDATE_ITEM = 'courses.updateItem';
export const COURSES_REPLACE_ITEM = 'courses.replaceItem';
export const COURSES_CREATE_ITEM = 'courses.createItem';

export const coursesPermissions = [
  COURSES_GET_ITEM,
  COURSES_GET_ITEMS,
  COURSES_DELETE_ITEM,
  COURSES_DELETE_ITEMS,
  COURSES_UPDATE_ITEM,
  COURSES_REPLACE_ITEM,
  COURSES_CREATE_ITEM,
];

export const ENROLMENTS_GET_ITEM = 'enrolments.getItem';
export const ENROLMENTS_GET_ITEMS = 'enrolments.getItems';
export const ENROLMENTS_DELETE_ITEM = 'enrolments.deleteItem';
export const ENROLMENTS_DELETE_ITEMS = 'enrolments.deleteItems';
export const ENROLMENTS_UPDATE_ITEM = 'enrolments.updateItem';
export const ENROLMENTS_REPLACE_ITEM = 'enrolments.replaceItem';
export const ENROLMENTS_CREATE_ITEM = 'enrolments.createItem';

export const enrolmentsPermissions = [
  ENROLMENTS_GET_ITEM,
  ENROLMENTS_GET_ITEMS,
  ENROLMENTS_DELETE_ITEM,
  ENROLMENTS_DELETE_ITEMS,
  ENROLMENTS_UPDATE_ITEM,
  ENROLMENTS_REPLACE_ITEM,
  ENROLMENTS_CREATE_ITEM,
];

export const basicUsersPermissions = [
  ENROLMENTS_GET_ITEM,
  ENROLMENTS_GET_ITEMS,
  USERS_GET_ITEM,
  USERS_GET_ITEMS,
  COURSES_GET_ITEM,
  COURSES_GET_ITEMS,
];

export const STUDENT_PERMISSIONS = [...basicUsersPermissions];

export const INSTRUCTOR_PERMISSIONS = [...basicUsersPermissions];

export const ADMIN_PERMISSIONS = [
  ...usersPermissions,
  ...enrolmentsPermissions,
  ...rolesPermissions,
  ...permissionsPermissions,
  ...coursesPermissions,
  // tslint:disable-next-line:max-file-line-count
];
