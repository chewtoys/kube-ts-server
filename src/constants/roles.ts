export const ADMIN = 'admin';
export const INSTRUCTOR = 'instructor';
export const STUDENT = 'student';
