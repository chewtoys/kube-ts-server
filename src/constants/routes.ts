export const API_V1 = '/api/v1';

export const AUTH = '/auth';
export const LOGIN = '/login';
export const REGISTER = '/register';
export const RESET_PASSWORD = '/reset-password';
export const REMIND_PASSWORD = '/remind-password';
export const VERIFY_ACCOUNT = '/verify-account';

export const USERS = '/users';

export const ROLES = '/roles';

export const PERMISSIONS = '/permissions';
