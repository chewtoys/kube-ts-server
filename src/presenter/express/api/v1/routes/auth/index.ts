import { Router } from 'express';
import {
  LOGIN,
  // REGISTER,
  // REMIND_PASSWORD,
  // RESET_PASSWORD,
} from '../../../../../../constants/routes';
import Config from '../../../../presenterFactory/Config';
import login from './login';
// import register from './register';
// import remindPassword from './remindPassword';
// import resetPassword from './resetPassword';

const authFactory = (config: Config): Router => {
  const router = Router();

  router.post(LOGIN, login(config));
  // router.post(REGISTER, register(config));
  // router.post(REMIND_PASSWORD, remindPassword(config));
  // router.post(RESET_PASSWORD, resetPassword(config));

  return router;
};

export default authFactory;
