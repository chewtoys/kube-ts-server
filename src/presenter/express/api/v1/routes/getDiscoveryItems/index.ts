import sendResponse from '@js-items/express/dist/utils/sendResponse';
import { toSnake } from 'convert-keys';
import { OK } from 'http-status-codes';
import Config from '../../../../presenterFactory/Config';
import catchErrors from '../../../../utils/errors/catchErrors';

const getDiscoveryItems = (config: Config) =>
  catchErrors(config, async (req, res) => {
    const type = req.query.type;
    
    let response;
    
    switch (type) {
      case 'homepage':
      default:
        response = await config.service.getDiscoveryItemsForHomepage({});
    }

    sendResponse({
      body: toSnake(response),
      req,
      res,
      status: OK,
    });
  });

export default getDiscoveryItems;
