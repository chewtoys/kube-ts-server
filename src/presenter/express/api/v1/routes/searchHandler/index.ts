import sendResponse from '@js-items/express/dist/utils/sendResponse';
import { toSnake } from 'convert-keys';
import { OK } from 'http-status-codes';
import Record from 'rulr/Record';
import validateData from 'rulr/validateData';
import { SAFE_URL_LENGTH } from '../../../../../../constants';
import String from '../../../../../../utils/validation/rules/String';
import Config from '../../../../presenterFactory/Config';
import catchErrors from '../../../../utils/errors/catchErrors';


const searchHandler = (config: Config) =>
  catchErrors(config, async (req, res) => {
    const name = config.appConfig.http.client.searchQueryParamName;

    const validationSchema = {
      [name]: String(0, SAFE_URL_LENGTH),
    };

    const rules = Record(validationSchema);
    const value = req.query[name];

    const payload: any = {[name]:value};

    validateData(rules)(payload);

    const response = await config.service.search({
      query: value
    });
       
    sendResponse({
      body: toSnake(response),
      req,
      res,
      status: OK,
    });
  });

export default searchHandler;
