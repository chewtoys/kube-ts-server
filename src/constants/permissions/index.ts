import { CATEGORIES_GET_ITEM, CATEGORIES_GET_ITEMS, catergoriesPermissions } from "./categories";
import { COURSES_GET_ITEM, COURSES_GET_ITEMS, coursesPermissions } from "./courses";
import { ENROLMENTS_GET_ITEM, ENROLMENTS_GET_ITEMS, enrolmentsPermissions } from "./enrolments";
import { permissionsPermissions } from "./permissions";
import { rolesPermissions } from "./roles";
import { USERS_GET_ITEM, USERS_GET_ITEMS, usersPermissions } from "./users";

export const basicUsersPermissions = [
  ENROLMENTS_GET_ITEM,
  ENROLMENTS_GET_ITEMS,
  USERS_GET_ITEM,
  USERS_GET_ITEMS,
  COURSES_GET_ITEM,
  COURSES_GET_ITEMS,
  CATEGORIES_GET_ITEM,
  CATEGORIES_GET_ITEMS,
];

export const STUDENT_PERMISSIONS = [...basicUsersPermissions];

export const INSTRUCTOR_PERMISSIONS = [...basicUsersPermissions];

export const ADMIN_PERMISSIONS = [
  ...usersPermissions,
  ...enrolmentsPermissions,
  ...rolesPermissions,
  ...permissionsPermissions,
  ...coursesPermissions,
  ...catergoriesPermissions,
  // tslint:disable-next-line:max-file-line-count
];
