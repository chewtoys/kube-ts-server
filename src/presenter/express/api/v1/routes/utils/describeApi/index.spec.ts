import dotenv from 'dotenv';
import assertOnResponseAndStatus from '../../../../../utils/tests/assertOnResponseAndStatus';
dotenv.config();
import { OK } from 'http-status-codes';
import { API_V1 } from '../../../../../../../constants/routes';
import initTests from '../../../../../utils/tests/initTests';

describe('@presenter/describeApi', () => {
  const { request } = initTests({ useMailServer: true });

  it('describes API metadata', async () => {
    await assertOnResponseAndStatus({
      method: 'get',
      request,
      statusCode: OK,
      url: API_V1,
    });
  });
});
