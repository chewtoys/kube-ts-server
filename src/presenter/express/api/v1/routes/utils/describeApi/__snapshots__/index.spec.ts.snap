// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@presenter/describeApi describes API metadata 1`] = `
Object {
  "available_routes": Object {
    "auth": Object {
      "login": "/api/v1/auth/login",
      "register": "/api/v1/auth/register",
      "remind_password": "/api/v1/auth/remind-password",
      "resend_verify_token": "/api/v1/auth/resend-verify-token",
      "reset_password": "/api/v1/auth/reset-password",
      "verify_account": "/api/v1/auth/verify-account",
    },
    "categories": "/api/v1/categories",
    "checks": Object {
      "liveness": "/health/liveness",
      "readiness": "/health/readiness",
      "version": "/version",
    },
    "courses": "/api/v1/courses",
    "discovery_items": "/api/v1/discovery-items",
    "enrolments": "/api/v1/enrolments",
    "permissions": "/api/v1/permissions",
    "roles": "/api/v1/roles",
    "users": "/api/v1/users",
  },
  "docs": "https://kubetsserver.docs.apiary.io",
  "issues": "https://github.com/kube-js/kube-ts-server/issues",
  "repo": "https://github.com/kube-js/kube-ts-server",
}
`;
